modmail/exts/utils/extensions.py

```diff
@@ -0,0 +1,282 @@
+            # This ensures the paginator will not cut off a page in the middle of a category.
+            category = category.replace("_", " ").title()
+            extensions = "\n".join(sorted(extensions))
+            lines.append(f"**{category}**\n{extensions}\n")
+
+        log.debug(f"{ctx.author} requested a list of all cogs. Returning a paginated list.")
+
+        # since we currently don't have a paginator.
+        output = ""
+        for line in lines:
+            output += line
+        await ctx.send(output)
+
+    def group_extension_statuses(self) -> t.Mapping[str, str]:
+        """Return a mapping of extension names and statuses to their categories."""
+        categories = {}
+        extensions = []
+        for ext, _nul in EXTENSIONS:
+            extensions.append(ext)
+        for ext in extensions:
+            if ext in self.bot.extensions:
+                status = ":green_circle:"
+            else:
+                status = ":red_circle:"
+
+            path = ext.split(".")
+            if len(path) > BASE_PATH_LEN + 1:
+                category = " - ".join(path[BASE_PATH_LEN:-1])
+            else:
+                category = "uncategorised"
+
+            categories.setdefault(category, []).append(f"{status}  {path[-1]}")
+
+        return categories
+
+    def batch_manage(self, action: Action, *extensions: str) -> str:
+        """
+        Apply an action to multiple extensions and return a message with the results.
+
+        If only one extension is given, it is deferred to `manage()`.
+        """
+        if len(extensions) == 1:
+            msg, _ = self.manage(action, extensions[0])
+            return msg
+
+        verb = action.name.lower()
+        failures = {}
+
+        for extension in extensions:
+            _, error = self.manage(action, extension)
+            if error:
+                failures[extension] = error
+
+        emoji = ":x:" if failures else ":ok_hand:"
+        msg = f"{emoji} {len(extensions) - len(failures)} / {len(extensions)} extensions {verb}ed."
+
+        if failures:
+            failures = "\n".join(f"{ext}\n    {err}" for ext, err in failures.items())
+            msg += f"\nFailures:```\n{failures}```"
+
+        log.debug(f"Batch {verb}ed extensions.")
+
+        return msg
+
+    def manage(self, action: Action, ext: str) -> t.Tuple[str, t.Optional[str]]:
+        """Apply an action to an extension and return the status message and any error message."""
+        verb = action.name.lower()
+        error_msg = None
+
+        try:
+            action.value(self.bot, ext)
+        except (commands.ExtensionAlreadyLoaded, commands.ExtensionNotLoaded):
+            if action is Action.RELOAD:
+                # When reloading, just load the extension if it was not loaded.
+                log.debug("Treating {ext!r} as if it was not loaded.")
+                return self.manage(Action.LOAD, ext)
```

Comment:
I don't think this hides anything.
![image](https://user-images.githubusercontent.com/71233171/128747165-90c37640-89a8-4141-bd7e-c467c16e7c72.png)

Reply:
What happens if an extension writes to a file on unload but crashes before that happens, and on load it attempts to read from the file? You'll get two errors on a reload, a failed unload, and then a failed load. We should instead fail early so this can be handled right rather than potentially damage extension state by performing two loads without an unload in between.

Reply:
Good point, I've removed the code that enabled this to happen.

The code now does not intentionally catch these errors, and sends a message if cogs are already loaded or not.